In [ ]:
# Подключаем библиотеки
import pandas as pd

**Задание 1.** <br>
Скачайте датасет fifаs2.сsv. Проанализируйте его и определите, какие данные являются неполными. Удалите
ненужные колонки и недостающие значения.

**Решение:**<br>

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/ETL/fifa_s2.csv', sep = ',')
df.head()

ID       Name  Age  Nationality  Overall  Potential  \
0   1179  G. Buffon   40        Italy       88         88   
1   5479   Casillas   37        Spain       82         82   
2   8205      Braga   35     Portugal       70         70   
3   9014  A. Robben   34  Netherlands       84         84   
4  19521  J. Ibehre   35      England       61         61   

                  Club    Value  Wage Preferred Foot  \
0  Paris Saint-Germain   4000.0    77          Right   
1             FC Porto   1500.0    10           Left   
2              CD Aves    750.0     5          Right   
3    FC Bayern München  15500.0   110           Left   
4     Cambridge United    120.0     2          Right   

   International Reputation  Skill Moves Position  Joined  \
0                       4.0          1.0       GK    2018   
1                       4.0          1.0      NaN    2015   
2                       1.0          3.0      LCM    2017   
3                       4.0          4.0       RM    2009   
4                       1.0          2.0       ST    2017   

  Contract Valid Until    Height  Weight  Release Clause  
0           01/01/2019  6.333333   203.0          7400.0  
1           01/01/2019  6.083333   185.0          3000.0  
2           01/01/2019  5.916667   165.0          1600.0  
3           01/01/2019  5.916667   176.0         25600.0  
4           01/01/2019  6.333333   203.0           210.0

In [ ]:
df.shape

(2399, 18)

Наименование полей:
- ID — идентификационный номер футболиста
- Name — имя
- Age — возраст
- Nationality — национальность
- Overall — в целом
- Potential — потенциал игрока (насколько высокого уровня он может достичь)
- Club — клуб футболиста
- Value — значение в евро
- Wage — заработная плата
- Preferred Foot — предпочтительная нога
- International Reputation — международный рейтинг
- Skill Moves — владение навыками
- Position — позиция игрока на поле
- Joined — год вступления в клуб
- Contract Valid Until — дата окончания контракта
- Height — рост игрока
- Weight — вес игрока
- Release Clause — освобождение от обязательств по контракту (отступные)

**Первичный анализ данных**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2399 entries, 0 to 2398
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        2399 non-null   int64  
 1   Name                      2399 non-null   object 
 2   Age                       2399 non-null   int64  
 3   Nationality               2392 non-null   object 
 4   Overall                   2399 non-null   int64  
 5   Potential                 2399 non-null   int64  
 6   Club                      2158 non-null   object 
 7   Value                     2142 non-null   float64
 8   Wage                      2399 non-null   int64  
 9   Preferred Foot            2399 non-null   object 
 10  International Reputation  2389 non-null   float64
 11  Skill Moves               2393 non-null   float64
 12  Position                  2378 non-null   object 
 13  Joined                    2399 non-null   int64  
 14  Contract

In [ ]:
df.isna().sum() # проверим наличие пустых значений в DataFrame

ID                            0
Name                          0
Age                           0
Nationality                   7
Overall                       0
Potential                     0
Club                        241
Value                       257
Wage                          0
Preferred Foot                0
International Reputation     10
Skill Moves                   6
Position                     21
Joined                        0
Contract Valid Until        273
Height                        0
Weight                        0
Release Clause                3
dtype: int64

**Вывод №1**. Представленные данные представляют таблицу, состояющую из 2399 строк и 18 столбцов.<br> Типы данных в таблице: float64(6), int64(6), object(6). <br> Пропущенные ячейки в таблице есть, в следующих столбцах:
1. Nationality - 7 ячеек
2. Club - 241 ячейка  
3. Value - 257 ячеек   
4. International Reputation - 10 ячеек
5. Skill Moves - 6 ячеек
6. Position - 21 ячейка
7. Contract Valid Until - 273 ячейки   
8. Release Clause - 3 ячейки.

In [ ]:
df.describe() # основные статистические характеристики данных по каждому числовому признаку

ID          Age      Overall    Potential          Value  \
count    2399.000000  2399.000000  2399.000000  2399.000000    2142.000000   
mean   212033.680700    25.213005    69.115048    74.115048    5980.873016   
std     30549.399068     4.633622     8.060195     7.190957   11963.420681   
min      1179.000000    16.000000    47.000000    48.000000      20.000000   
25%    195105.000000    22.000000    64.000000    69.000000     500.000000   
50%    218745.000000    25.000000    69.000000    74.000000    1100.000000   
75%    234576.000000    29.000000    74.000000    79.000000    6000.000000   
max    246608.000000    44.000000    94.000000    95.000000  118500.000000   

              Wage  International Reputation  Skill Moves       Joined  \
count  2399.000000               2389.000000  2393.000000  2399.000000   
mean     24.178825                  1.308497     2.469703  2016.145894   
std      49.552098                  0.687175     0.844531     2.085829   
min       0.000000                  1.000000     1.000000  2000.000000   
25%       1.000000                  1.000000     2.000000  2016.000000   
50%       5.000000                  1.000000     2.000000  2017.000000   
75%      22.000000                  1.000000     3.000000  2018.000000   
max     565.000000                  5.000000     5.000000  2018.000000   

            Height       Weight  Release Clause  
count  2399.000000  2399.000000     2396.000000  
mean      5.955300   166.802782    10690.033498  
std       0.219249    15.241015    21780.029323  
min       5.250000   121.000000       35.000000  
25%       5.833333   154.000000     1100.000000  
50%       6.000000   165.000000     3800.000000  
75%       6.083333   176.000000     7500.000000  
max       6.583333   214.000000   228100.000000

**Вывод №2**. Согласно полученным статистическим характеристикам средние значения по игрокам следующие:<br>
- Age - 25 лет
- Overall - 69
- Potential -74
- Value - 5980
- Wage - 24
- International Reputation - 1
- Skill Moves - 2
- Joined - 2016
- Height - 5,95 футов
- Weight - 166,8 фунтов
- Release Clause -10 690
Но согласно полученным общим статистическим характеристикам по игрокам, видно, что значения в столбцах, Value, Wage, Release Clause (50% сильно отличаются от mean) являются ненормальными.


**Выполним очистку данных**

In [ ]:
# Нулевые качественные значения NaN в столбцах заменим на строку 'none'
df[['Position', 'Nationality', 'Club', 'Contract Valid Until']] = df[['Position', 'Nationality', 'Club', 'Contract Valid Until']]. fillna('none')

# Количественные значения равные 0 (241), заполним mean значением
df['Wage'].replace(0,df['Wage'].mean(),inplace=True)

# Нулевые количественные значения NaN(257) в столбце Value заменим на значение mean
df['Value'] = df['Value']. fillna(df['Value'].mean())
df[['Release Clause']] = df[['Release Clause']]. fillna (35)

# Нулевые количественные значения NaN в столбцах заменим на значение min 1
df[['Skill Moves', 'International Reputation',]] = df[['Skill Moves', 'International Reputation']]. fillna (1)

In [ ]:
df.isna().sum() # проверим наличие пустых значений в DataFrame после очистки данных

ID                          0
Name                        0
Age                         0
Nationality                 0
Overall                     0
Potential                   0
Club                        0
Value                       0
Wage                        0
Preferred Foot              0
International Reputation    0
Skill Moves                 0
Position                    0
Joined                      0
Contract Valid Until        0
Height                      0
Weight                      0
Release Clause              0
dtype: int64

**Удалим не значимые столбцы в DataFrame**

In [ ]:
df = df.drop(['Joined','Contract Valid Until'],axis='columns')               # Удалить нерелевантные столбцы
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2399 entries, 0 to 2398
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        2399 non-null   int64  
 1   Name                      2399 non-null   object 
 2   Age                       2399 non-null   int64  
 3   Nationality               2399 non-null   object 
 4   Overall                   2399 non-null   int64  
 5   Potential                 2399 non-null   int64  
 6   Club                      2399 non-null   object 
 7   Value                     2399 non-null   float64
 8   Wage                      2399 non-null   float64
 9   Preferred Foot            2399 non-null   object 
 10  International Reputation  2399 non-null   float64
 11  Skill Moves               2399 non-null   float64
 12  Position                  2399 non-null   object 
 13  Height                    2399 non-null   float64
 14  Weight  

**Задание 2.** <br>
Найдите в датафрейме полные дубликаты и удалите их. Значения могут быть одинаковыми, но написаны по-разному. Например, может отличаться размер регистра (заглавные и строчные буквы). Особое внимание уделить колонке с названиями команд.

**Решение:**<br>

In [ ]:
# Приведем строковые значения столбцов в нижний регистр
df['Name'] = df['Name'].str.lower()
df['Nationality'] = df['Nationality'].str.lower()
df['Club'] = df['Club'].str.lower()
df['Preferred Foot'] = df['Preferred Foot'].str.lower()
df['Position'] = df['Position'].str.lower()
df

ID           Name  Age  Nationality  Overall  Potential  \
0       1179      g. buffon   40        italy       88         88   
1       5479       casillas   37        spain       82         82   
2       8205          braga   35     portugal       70         70   
3       9014      a. robben   34  netherlands       84         84   
4      19521      j. ibehre   35      england       61         61   
...      ...            ...  ...          ...      ...        ...   
2394  246465       guerrero   16        spain       63         76   
2395  246594    h. massengo   16       france       62         75   
2396  246599  guillem jaime   19        spain       64         80   
2397  246606    fran garcía   18        spain       64         79   
2398  246608        fidalgo   21        spain       65         75   

                     Club    Value   Wage Preferred Foot  \
0     paris saint-germain   4000.0   77.0          right   
1                fc porto   1500.0   10.0           left   
2                 cd aves    750.0    5.0          right   
3       fc bayern münchen  15500.0  110.0           left   
4        cambridge united    120.0    2.0          right   
...                   ...      ...    ...            ...   
2394  cf rayo majadahonda    625.0    1.0           left   
2395            as monaco    450.0    2.0          right   
2396         fc barcelona    975.0   13.0          right   
2397          real madrid    825.0    9.0           left   
2398          real madrid    875.0   20.0          right   

      International Reputation  Skill Moves Position    Height  Weight  \
0                          4.0          1.0       gk  6.333333   203.0   
1                          4.0          1.0     none  6.083333   185.0   
2                          1.0          3.0      lcm  5.916667   165.0   
3                          4.0          4.0       rm  5.916667   176.0   
4                          1.0          2.0       st  6.333333   203.0   
...                        ...          ...      ...       ...     ...   
2394                       1.0          3.0       cm  5.916667   165.0   
2395                       1.0          2.0      cdm  5.833333   146.0   
2396                       1.0          3.0       rw  5.833333   154.0   
2397                       1.0          2.0       lb  5.583333   139.0   
2398                       1.0          3.0       cm  5.750000   150.0   

      Release Clause  
0             7400.0  
1             3000.0  
2             1600.0  
3            25600.0  
4              210.0  
...              ...  
2394          1400.0  
2395          1100.0  
2396          2600.0  
2397          2200.0  
2398          2000.0  

[2399 rows x 16 columns]

В названиях клубов имеются значения (1.), пробелы, пропуски, преходы.

In [ ]:
# Удаляем пробелы и переходы на новую строку.
df['Name'] = df['Name'].str.replace('\n', '')
df['Name'] = df['Name'].str.replace('\t', ' ')
df['Name'] = df['Name'].str.replace(' {2,}', ' ', regex=True)
df['Name'] = df['Name'].str.strip()

df['Club'] = df['Club'].str.replace('1.', '', regex=True) # удаляем (1.) в названиях клубов
df['Club'] = df['Club'].str.replace('\n', '')
df['Club'] = df['Club'].str.replace('\t', ' ')
df['Club'] = df['Club'].str.replace(' {2,}', ' ', regex=True)
df['Club'] = df['Club'].str.strip()
df

In [ ]:
df.nunique() # количество уникальных наименований

ID                          2395
Name                        2371
Age                           27
Nationality                  103
Overall                       47
Potential                     44
Club                          79
Value                        206
Wage                         141
Preferred Foot                 2
International Reputation       5
Skill Moves                    5
Position                      28
Height                        18
Weight                        43
Release Clause               715
dtype: int64

**Найдем полные дубликаты**

**1. Проверим столбец 'ID', так как  из 2399 строк 2395 уникальны**

In [ ]:
df[df.ID.duplicated()]

ID        Name  Age Nationality  Overall  Potential  \
154   168651  i. rakitić   30     croatia       87         87   
304   181872    a. vidal   31       chile       85         85   
687   200466   v. eicher   27     germany       67         69   
2123  241736  y. bisseck   17     germany       61         82   

                  Club    Value   Wage Preferred Foot  \
154       fc barcelona  46500.0  260.0          right   
304       fc barcelona  26000.0  205.0          right   
687   fc heidenheim 46    625.0    4.0           left   
2123           fc köln    500.0    1.0          right   

      International Reputation  Skill Moves Position    Height  Weight  \
154                        4.0          3.0      rcm  6.000000   172.0   
304                        4.0          3.0      cam  5.916667   165.0   
687                        1.0          1.0       gk  6.250000   187.0   
2123                       1.0          2.0       cb  6.333333   183.0   

      Release Clause  
154          95300.0  
304          53300.0  
687           1000.0  
2123          1300.0

In [ ]:
df[df['ID'] == 168651] # проверка дубликата

ID        Name  Age Nationality  Overall  Potential          Club  \
153  168651  i. rakitić   30     croatia       87         87  fc barcelona   
154  168651  i. rakitić   30     croatia       87         87  fc barcelona   

       Value   Wage Preferred Foot  International Reputation  Skill Moves  \
153  46500.0  260.0          right                       4.0          3.0   
154  46500.0  260.0          right                       4.0          3.0   

    Position  Height  Weight  Release Clause  
153      rcm     6.0   172.0         95300.0  
154      rcm     6.0   172.0         95300.0

In [ ]:
df[df['ID'] == 181872] # проверка дубликата

ID      Name  Age Nationality  Overall  Potential          Club  \
303  181872  a. vidal   31       chile       85         85  fc barcelona   
304  181872  a. vidal   31       chile       85         85  fc barcelona   

       Value   Wage Preferred Foot  International Reputation  Skill Moves  \
303  26000.0  205.0          right                       4.0          3.0   
304  26000.0  205.0          right                       4.0          3.0   

    Position    Height  Weight  Release Clause  
303      cam  5.916667   165.0         53300.0  
304      cam  5.916667   165.0         53300.0

In [ ]:
df[df['ID'] == 200466] # проверка дубликата

ID       Name  Age Nationality  Overall  Potential              Club  \
686  200466  v. eicher   27     germany       67         69  fc heidenheim 46   
687  200466  v. eicher   27     germany       67         69  fc heidenheim 46   

     Value  Wage Preferred Foot  International Reputation  Skill Moves  \
686  625.0   4.0           left                       1.0          1.0   
687  625.0   4.0           left                       1.0          1.0   

    Position  Height  Weight  Release Clause  
686       gk    6.25   187.0          1000.0  
687       gk    6.25   187.0          1000.0

In [ ]:
df[df['ID'] == 241736] # проверка дубликата

ID        Name  Age Nationality  Overall  Potential     Club  Value  \
2122  241736  y. bisseck   17     germany       61         82  fc köln  500.0   
2123  241736  y. bisseck   17     germany       61         82  fc köln  500.0   

      Wage Preferred Foot  International Reputation  Skill Moves Position  \
2122   1.0          right                       1.0          2.0       cb   
2123   1.0          right                       1.0          2.0       cb   

        Height  Weight  Release Clause  
2122  6.333333   183.0          1300.0  
2123  6.333333   183.0          1300.0

**Вывод:** Как видно по столбцу 'ID' найдено 4 полных дубликата.

In [ ]:
# Удалим полные дубликаты
df = df.drop_duplicates(subset=['ID'], keep='last')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2395 entries, 0 to 2398
Data columns (total 16 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   ID                        2395 non-null   int64  
 1   Name                      2395 non-null   object 
 2   Age                       2395 non-null   int64  
 3   Nationality               2395 non-null   object 
 4   Overall                   2395 non-null   int64  
 5   Potential                 2395 non-null   int64  
 6   Club                      2395 non-null   object 
 7   Value                     2395 non-null   float64
 8   Wage                      2395 non-null   float64
 9   Preferred Foot            2395 non-null   object 
 10  International Reputation  2395 non-null   float64
 11  Skill Moves               2395 non-null   float64
 12  Position                  2395 non-null   object 
 13  Height                    2395 non-null   float64
 14  Weight  

In [ ]:
df.ID.duplicated().sum() # проверим дубликаты по столбцу 'ID'


0

**2. Проверим столбец 'Name', так как из 2395 строк 2371 уникальны**

In [ ]:
df.nunique() # количество уникальных наименований

ID                          2395
Name                        2371
Age                           27
Nationality                  103
Overall                       47
Potential                     44
Club                          79
Value                        206
Wage                         141
Preferred Foot                 2
International Reputation       5
Skill Moves                    5
Position                      28
Height                        18
Weight                        43
Release Clause               715
dtype: int64

In [ ]:
df[df.Name.duplicated()]

ID           Name  Age Nationality  Overall  Potential  \
191   172680    s. sinclair   30     england       60         60   
593   194730    juan carlos   30       spain       71         71   
641   198710   j. rodríguez   26    colombia       88         89   
719   201400        rafinha   25      brazil       82         87   
783   203668  a. donnarumma   27       italy       75         75   
857   206083      j. murphy   23     england       70         77   
892   207863         felipe   29      brazil       83         83   
1248  220540    d. bejarano   24     bolivia       63         66   
1250  220605   j. rodríguez   22    colombia       66         76   
1294  222368   a. miranchuk   22      russia       74         81   
1441  225563     joão pedro   21      brazil       72         81   
1461  225900   j. sambenito   26    paraguay       71         74   
1556  229026   s. abdullahi   25     nigeria       72         75   
1730  232805       bernardo   23      brazil       75         80   
1907  237156       m. wagué   19     senegal       69         81   
1908  237160    r. williams   19     england       64         79   
1979  238789     h. herrera   20       chile       56         66   
1985  238877         bastos   26       spain       66         69   
2018  239928      t. müller   23     germany       65         69   
2076  240915        miranda   18       spain       66         81   
2110  241559   c. gutiérrez   18       chile       52         70   
2114  241597       b. white   26   australia       56         59   
2203  243237  l. pellegrini   19       italy       65         79   
2394  246465       guerrero   16       spain       63         76   

                        Club         Value        Wage Preferred Foot  \
191           bristol rovers    190.000000    2.000000          right   
593                  cd lugo   1500.000000    5.000000          right   
641        fc bayern münchen  69500.000000  315.000000           left   
719             fc barcelona  29500.000000  165.000000           left   
783                  brescia   7000.000000    6.000000          right   
857             cardiff city   2500.000000   21.000000          right   
892                 fc porto  20000.000000   22.000000          right   
1248                    none   5980.873016   24.178825           left   
1250             cd tenerife   1000.000000    3.000000          right   
1294        lokomotiv moscow   8000.000000    1.000000          right   
1441                fc porto   4100.000000    6.000000          right   
1461                    none   5980.873016   24.178825          right   
1556               bursaspor   3300.000000   12.000000          right   
1730  brighton & hove albion   7500.000000   33.000000           left   
1907            fc barcelona   1600.000000   20.000000          right   
1908       manchester united    875.000000    8.000000          right   
1979            cd o'higgins    140.000000    1.000000          right   
1985        cf reus deportiu    650.000000    3.000000          right   
2018            fc magdeburg    600.000000    3.000000          right   
2076            fc barcelona   1200.000000   12.000000           left   
2110          cd antofagasta    100.000000    1.000000          right   
2114           brisbane roar     80.000000    1.000000          right   
2203                    roma   1000.000000    5.000000           left   
2394     cf rayo majadahonda    625.000000    1.000000           left   

      International Reputation  Skill Moves Position    Height  Weight  \
191                        1.0          2.0       cm  5.583333   148.0   
593                        1.0          1.0       gk  6.166667   181.0   
641                        4.0          4.0      lam  5.916667   172.0   
719                        2.0          4.0      cam  5.750000   157.0   
783                        1.0          3.0       ls  5.916667   165.0   
857                        1.0         

In [ ]:
# Проверка дубликатов
df[df['Name'] == 's. sinclair']

ID         Name  Age Nationality  Overall  Potential            Club  \
156  169214  s. sinclair   29     england       75         75          celtic   
191  172680  s. sinclair   30     england       60         60  bristol rovers   

      Value  Wage Preferred Foot  International Reputation  Skill Moves  \
156  6500.0  48.0          right                       2.0          4.0   
191   190.0   2.0          right                       1.0          2.0   

    Position    Height  Weight  Release Clause  
156       lm  5.833333   154.0         11700.0  
191       cm  5.583333   148.0           333.0

**Вывод:** Как видно по столбцу 'Name' дубликаты показали одинаковые имена, но  разные ID, люди являются однофамильцами.

**Задание 3.** <br>
Напишите функцию, которая добавит колонку с разбиением возраста по группам: до 20, от 20 до 30, от 30 до 36 и старше 36. Посчитайте количество футболистов в каждой категории.
ЗАДАНИЕ НУЖНО СДЕЛАТЬ С ПОМОЧЬЮ APACHE SPARK КАК ПОКАЗАНО НА СЕМИНАРЕ. Соберите все в один pdf, код и скриншоты записанных таблиц с данными, скриншоты отработки лога спарка.

**Решение (ВАРИАНТ №1):**<br>

In [ ]:
df.loc[df['Age']<=20, 'Age_category'] = 'before 20 years'
df.loc[df['Age'].between(20,30), 'Age_category'] = '20-30 years'
df.loc[df['Age'].between(30,36), 'Age_category'] = '30-36 years'
df.loc[df['Age']>36, 'Age_category'] = 'after 36 years'
df

ID           Name  Age  Nationality  Overall  Potential  \
0       1179      g. buffon   40        italy       88         88   
1       5479       casillas   37        spain       82         82   
2       8205          braga   35     portugal       70         70   
3       9014      a. robben   34  netherlands       84         84   
4      19521      j. ibehre   35      england       61         61   
...      ...            ...  ...          ...      ...        ...   
2394  246465       guerrero   16        spain       63         76   
2395  246594    h. massengo   16       france       62         75   
2396  246599  guillem jaime   19        spain       64         80   
2397  246606    fran garcía   18        spain       64         79   
2398  246608        fidalgo   21        spain       65         75   

                     Club    Value   Wage Preferred Foot  \
0     paris saint-germain   4000.0   77.0          right   
1                fc porto   1500.0   10.0           left   
2                 cd aves    750.0    5.0          right   
3       fc bayern münchen  15500.0  110.0           left   
4        cambridge united    120.0    2.0          right   
...                   ...      ...    ...            ...   
2394  cf rayo majadahonda    625.0    1.0           left   
2395            as monaco    450.0    2.0          right   
2396         fc barcelona    975.0   13.0          right   
2397          real madrid    825.0    9.0           left   
2398          real madrid    875.0   20.0          right   

      International Reputation  Skill Moves Position    Height  Weight  \
0                          4.0          1.0       gk  6.333333   203.0   
1                          4.0          1.0     none  6.083333   185.0   
2                          1.0          3.0      lcm  5.916667   165.0   
3                          4.0          4.0       rm  5.916667   176.0   
4                          1.0          2.0       st  6.333333   203.0   
...                        ...          ...      ...       ...     ...   
2394                       1.0          3.0       cm  5.916667   165.0   
2395                       1.0          2.0      cdm  5.833333   146.0   
2396                       1.0          3.0       rw  5.833333   154.0   
2397                       1.0          2.0       lb  5.583333   139.0   
2398                       1.0          3.0       cm  5.750000   150.0   

      Release Clause     Age_category  
0             7400.0   after 36 years  
1             3000.0   after 36 years  
2             1600.0      30-36 years  
3            25600.0      30-36 years  
4              210.0      30-36 years  
...              ...              ...  
2394          1400.0  before 20 years  
2395          1100.0  before 20 years  
2396          2600.0  before 20 years  
2397          2200.0  before 20 years  
2398          2000.0      20-30 years  

[2395 rows x 17 columns]

In [ ]:
# Количество футболистов в каждой возрастной категории
df['Age_category'].value_counts()

20-30 years        1645
30-36 years         466
before 20 years     269
after 36 years       15
Name: Age_category, dtype: int64

**Решение (ВАРИАНТ №2) c помощью APACHE SPARK приложен в pdf**<br>